akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -38,92 +39,175 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   def flowController: IncomingFlowController = IncomingFlowController.default(settings)
 
-  private var incomingStreams = new immutable.TreeMap[Int, IncomingStreamState]
+  private var streamStates = new immutable.TreeMap[Int, StreamState]
   private var largestIncomingStreamId = 0
   private var outstandingConnectionLevelWindow = Http2Protocol.InitialWindowSize
   private var totalBufferedData = 0
 
-  private def streamFor(streamId: Int): IncomingStreamState =
-    incomingStreams.get(streamId) match {
+  /**
+   * The "last peer-initiated stream that was or might be processed on the sending endpoint in this connection"
+   *
+   * @see http://httpwg.org/specs/rfc7540.html#rfc.section.6.8
+   */
+  def lastStreamId(): Int = largestIncomingStreamId
+
+  private def streamFor(streamId: Int): StreamState =
+    streamStates.get(streamId) match {
       case Some(state) => state
       case None =>
         if (streamId <= largestIncomingStreamId) Closed // closed streams are never put into the map
         else if (isUpgraded && streamId == 1) {
+          require(isServer)
           // Stream 1 is implicitly "half-closed" from the client toward the server (see Section 5.1), since the request is completed as an HTTP/1.1 request
           // https://http2.github.io/http2-spec/#discover-http
           largestIncomingStreamId = streamId
-          incomingStreams += streamId -> HalfClosedRemote
-          HalfClosedRemote
+          streamStates += streamId -> HalfClosedRemoteWaitingForOutgoingStream(0)
+          HalfClosedRemoteWaitingForOutgoingStream(0)
         } else {
           largestIncomingStreamId = streamId
-          incomingStreams += streamId -> Idle
+          streamStates += streamId -> Idle // FIXME: is that needed at all? Should move out of Idle immediately
           Idle
         }
     }
 
-  def handleStreamEvent(e: StreamFrameEvent): Unit = {
+  /** Called by Http2ServerDemux to let the state machine handle StreamFrameEvents */
+  def handleStreamEvent(e: StreamFrameEvent): Unit =
     updateState(e.streamId, _.handle(e))
-  }
 
+  /** Called by Http2ServerDemux when a stream comes in from the user-handler */
   def handleOutgoingCreated(stream: Http2SubStream): Unit =
-    updateState(stream.streamId, _.handleOutgoingCreated(stream))
+    if (streamFor(stream.streamId) != Closed) {
+      multiplexer.pushControlFrame(stream.initialHeaders)
+
+      if (stream.initialHeaders.endStream) {
+        updateState(stream.streamId, _.handleOutgoingCreatedAndFinished(stream.correlationAttributes))
+      } else {
+        val outStream = OutStream(stream)
+        stream.initialHeaders.priorityInfo.foreach(multiplexer.updatePriority)
```

Comment:
Shouldn't we keep this information even for requests without an entity? so when other streams depend on this stream, we can correctly infer that they should depend on the stream that this stream depended on?

Reply:
I haven't totally digested https://tools.ietf.org/html/rfc7540#section-5.3.4 yet, but noticed:

```
   It is possible for a stream to become closed while prioritization
   information that creates a dependency on that stream is in transit.
   If a stream identified in a dependency has no associated priority
   information, then the dependent stream is instead assigned a default
   priority (Section 5.3.5).  This potentially creates suboptimal
   prioritization, since the stream could be given a priority that is
   different from what is intended.

   To avoid these problems, an endpoint SHOULD retain stream
   prioritization state for a period after streams become closed
```

Reply:
I pulled that statement out, so that the multiplexer gets to know about priority changes in all cases (even the Closed case).

Right now it doesn't matter because we don't have a prioritization strategy that uses any of the priority information.